metomi/isodatetime/data.py

```diff
@@ -522,54 +542,50 @@ def __floordiv__(self, other):
                 "'%s' should be integer." %
                 type(other).__name__
             )
-        new = self.copy()
+        new = self._copy()
         if self.get_is_in_weeks():
-            new.weeks //= other
+            new._weeks //= other
             return new
-        new.years //= other
-        new.months //= other
-        new.days //= other
-        new.hours //= other
-        new.minutes //= other
-        new.seconds //= other
+        new._years //= other
+        new._months //= other
+        new._days //= other
+        new._hours //= other
+        new._minutes //= other
+        new._seconds //= other
         return new
 
+    def __hash__(self) -> int:
+        return hash(self.get_days_and_seconds())
```

Comment:
Because `get_days_and_seconds` duplicates the instance it would probably be more efficient to extract the required data into a tuple.

```python
return hash((getattr(self, attr) for attr in self.DATA_ATTRIBUTES))
```

Reply:
The problem then is it would mean `PT1H != PT60M`

Reply:
I think for a `__hash__` that is the correct answer.

Reply:
However I have just spotted ways that `get_days_and_seconds()` can be optimised

Reply:
46eb487 makes it 10x faster when the Duration is in weeks, but can't really speed up otherwise

Reply:
> I think for a __hash__ that is the correct answer.

According to [the docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__):
> The only required property is that objects which compare equal have the same hash value

Reply:
Having two instances with different attributes producing the same hash kinda breaks the contract of `__hash__`.

We can't have `P1Y` and `P365D` producing the same hash as this will cause issues with caching logic e.g:

```python
# untested

from functools import lru_cache

@lru_cahe
def add(this, that):
    return this + that

t1 = TimePoint('2000')
d1 = Duration('P1Y')
d2 = Duration('P365D') 

# do this one way
add(t1, d1)  # 2001-01-01
add(t1, d2)  # 2001-01-01

# but do it the other
add.clear()
add(t1, d2)  # 2000-12-31
add(t1, d1)  # 2000-12-31
```

The nicest way out of this might be to change the behaviour of `__eq__`, I'm trying to work out what edge cases this actually applies to:

```python
def __eq__(self, other):
    if (
            # cases for which get_days_and_seconds is not accurate
            self.years != other.years
            and self.months != other.months
            # ...
    ):
        return False
    return self.get_days_and_seconds() == other.get_days_and_seconds()
```

Reply:
I don't have an answer yet, but having revisited the rationale behind doing this, I think the main criterion for the hash being the same has to be whether the methods will return the same results.

And I agree that `x` years or months can't be considered the same as `y` days

(except for the case of the 360-day calendar, and years in the 365/6-day calendars, ugh)

Reply:
> ugh

🤢

Reply:
Another thing: `P1D` is not necessarily `PT24H`! Because there might be a daylight saving change overnight

Reply:
Good point, so why did we try to determine equality.

Reply:
At least in the case of hours, minutes & seconds, they are always exact and equality definitely applies for them (having reviewed the 8601 standard - [draft spec here](https://www.loc.gov/standards/datetime/iso-tc154-wg5_n0038_iso_wd_8601-1_2016-02-16.pdf))